### ISSUE TITLE
Active ding video

### ISSUE THREAD

**Issue by DustinBryant** [dgreif/ring#117]
@dgreif, i'd like to thank you tremendously on your endeavors getting this solid Ring API together. The work you've done has gotten me very close to creating a module for my Magic Mirror (https://github.com/MichMich/MagicMirror) home project.

Excuse me if my terminology is wrong here. 

Is it possible to get an active ding video information similar to how you handle vod? As in, getting stream.m3u8 data and video (.ts) sent from ring?

I'm combining both your example and browser example to come up with a solution here. I've been able to get them to work independently within the magic mirror module I'm working on. However, I'm struggling to find a way to use the example when a ding is triggered to then use that data to send to ring to get similar information as you get back with your browser example without creating a new vod event.

Basically, if my ring doorbell has received a ring, I'd like to then get that live video instantly and display it. I feel like all the mechanisms are there I'm just missing something.

Thanks again!

**Comment by dgreif** [comment]
@DustinBryant glad to see the api integrated into a cool project!  You aren't missing anything, the mechanism you need doesn't exist yet.  Currently the only convenient way to start a video stream creates a new VOD ding, regardless of what active dings already exists.  I did this to avoid cases where the user might start streaming from an active ding via the Ring app, which would cause a second stream connecting with the same sip credentials to fail.  You could _technically_ make it work by creating a new instance of the `SipSession` manually, but you would have to copy everything that is being done inside the `createSipSession` method.  I can think of two potential solutions on my end:

1. I could allow `createSipSession` to take an `ActiveDing` as a second (optional) parameter.  This way you could give me the exact ding that you want to use to set up the `SipSession`
2. I could make a second (optional) boolean parameter that gives you the option to use any existing `ActiveDing`s that we already know about.  These stick around in memory for ~1 minute after the event occurs.  If we went this route, you wouldn't need to think about which ding to use, it would just grab one if available, or create a new VOD ding if there aren't any active dings within the last minute

I'm open to any other suggestions you have, and your feedback on these two options.

**Comment by DustinBryant** [comment]
I didn't even think to check it yet but is it even possible to start a VOD while there is an ActiveDing occurring? I wouldn't be opposed to just doing that, I just assumed there couldn't be two events on the same camera happening at the same time (doorbell manually being rang and a VOD session).

I definitely want to avoid not being able to still use the Ring app to answer the ActiveDing.

If it isn't possible to have two active events on the same camera, if you went with either of the options you provided then this functionality would error the Ring app, correct (assuming here)?

If we can get the active ding info without error-ing out the Ring app, then with the first option, I'm assuming the SipSession would produce the file outputs just like the VOD?

As for the second option, what call are you referring to that would get the optional bool parameter (we still talking createSipSession())? When you say they stick around in memory for a 1 minute, do you mean we'd only get a 1 minute feed before it expires?

Sorry, this whole sip and video stream is new territory for me so I may not be seeing the obvious here.

**Comment by dgreif** [comment]
For now, I'd say just use the standard VOD setup as it is and don't worry about the duplicate events.  I'm going to do some testing on my end and figure out how to do simultaneous streams.  I currently have a bug where I get a `480` SIP error if I connect multiple streams or reconnect too quickly to the same camera.  The Ring apps can stream simultaneously from the camera, so I must be doing something slightly wrong.  Will push an update whenever I find a fix (probably some time in the next week)

**Comment by DustinBryant** [comment]
Sounds good! I'll report back with my findings using that approach. 

Again, thank you for doing all the heavy lifting here and reverse engineering this bad boy.

**Comment by DustinBryant** [comment]
Ok, so I have it working using the standard VOD setup. Once the doorbell is pressed I clear out the output folder and put on chokidar (file watcher) waiting for the m3u8 file to appear. Once the file appears I hook it up to hls and the live video appears.

What I've noticed is that it doesn't start a new event, but instead hooks into the current ring event and starts downloading the stream in pieces. The only issue with it (which I may just be able to overlook), is that it will "answer" the event automatically when doing this approach. Which I gather you were already aware of.

That's really the only limitation I'm noticing as of right now. Other than that, it works flawlessly.

**Comment by dgreif** [comment]
@DustinBryant glad to hear it's working for you!  I think what Ring does is mark a ding or motion event as "answered" whenever a live stream is created for that ding event.  I'm still playing around with the SIP setup to ensure we don't break the Ring app, but I think they are doing something slightly different that allows the app to do multiple streams from the same event.  I'm missing some small detail, but hopefully I will get it tracked down this week.  I also confirmed on my end that they reuse ding events if you try to create a new VOD and one is still active.  Let me know if you run into any other issues with the current setup.

**Comment by DustinBryant** [comment]
@dgreif sorry to throw this one at you but I'm kind of lost on my end with an issue being reported by others. If you can't be bothered I totally get it.

This is in relation to showing the video stream via hls.js. I'm not sure if this is an encoding issue or an hls.js issue. I figured you had experience with ffmpeg and hls.js since your examples use them. Some users report the video doesn't appear when it should.

The basic setup is when a camera receives onDoorbellPressed I start a session. I do this by passing in the same options as in your browser example into camera.streamVideo().

```js
		this.sipSession = await camera.streamVideo({
			output: [
				"-preset",
				"veryfast",
				"-g",
				"25",
				"-sc_threshold",
				"0",
				"-f",
				"hls",
				"-hls_time",
				"2",
				"-hls_list_size",
				"6",
				"-hls_flags",
				"delete_segments",
				pathApi.join(this.videoOutputDirectory, this.audioPlaylistFile)
			]
		});
```
I then put on a file watcher that once the m3u8 appears in the designated this.videoOutputDirectory I then begin hooking up hls.js to a video element and providing it the m3u8 path. Once the MANIFEST_PARSED occurs I do a video.play() just like your browser example.

Now, I'm not experiencing this issue on my end yet I've had a couple people tell me they are. Is there some ffmpeg configuration or hls option that could assist here? Perhaps you've run into something similar?

For reference the code for this can be found here: https://github.com/DustinBryant/MMM-Ring

The ffmpeg encoding can be found in node_helper.js and the hls/html can be found in MMM-Ring.js (under case this.DisplayTypes.VIDEO:)

**Comment by dgreif** [comment]
@DustinBryant sorry if you are still stuck on this issue.  Unfortunately I don't know anything about m3u8 (that example was provided by @bourdakos1).  Good luck tracking it down.  Let me know if you find any issues with the api!
